pkg/oc/cli/cmd/image/mirror/s3.go

```diff
@@ -0,0 +1,454 @@
+}
+
+type s3CredentialStore struct {
+	store     auth.CredentialStore
+	url       *url.URL
+	retrieved bool
+}
+
+func (s *s3CredentialStore) IsExpired() bool { return !s.retrieved }
+
+func (s *s3CredentialStore) Retrieve() (credentials.Value, error) {
+	s.retrieved = false
+	accessKeyID, secretAccessKey := s.store.Basic(s.url)
+	if len(accessKeyID) == 0 || len(secretAccessKey) == 0 {
+		return credentials.Value{}, fmt.Errorf("no AWS credentials located for %s", s.url)
+	}
+	s.retrieved = true
+	glog.V(4).Infof("found credentials for %s", s.url)
+	return credentials.Value{
+		AccessKeyID:     accessKeyID,
+		SecretAccessKey: secretAccessKey,
+		ProviderName:    "DockerCfg",
+	}, nil
+}
+
+func (d *s3Driver) newObject(server *url.URL, region string, insecure bool, securityDomain *url.URL) (*s3.S3, error) {
+	key := fmt.Sprintf("%s:%s:%t:%s", server, region, insecure, securityDomain)
+	s3obj, ok := d.repositories[key]
+	if ok {
+		return s3obj, nil
+	}
+
+	awsConfig := aws.NewConfig()
+
+	var creds *credentials.Credentials
+	creds = credentials.NewChainCredentials([]credentials.Provider{
+		&s3CredentialStore{store: d.Creds, url: securityDomain},
+		&credentials.EnvProvider{},
+	})
+
+	awsConfig.WithS3ForcePathStyle(true)
+	awsConfig.WithEndpoint(server.String())
+	awsConfig.WithCredentials(creds)
+	awsConfig.WithRegion(region)
+	awsConfig.WithDisableSSL(insecure)
+	if glog.V(6) {
+		awsConfig.WithLogLevel(aws.LogDebug)
+	}
+
+	if d.UserAgent != "" {
+		awsConfig.WithHTTPClient(&http.Client{
+			Transport: transport.NewTransport(http.DefaultTransport, transport.NewHeaderRequestModifier(http.Header{http.CanonicalHeaderKey("User-Agent"): []string{d.UserAgent}})),
+		})
+	}
+	s, err := session.NewSession(awsConfig)
+	if err != nil {
+		return nil, err
+	}
+	s3obj = s3.New(s)
+	if d.repositories == nil {
+		d.repositories = make(map[string]*s3.S3)
+	}
+	d.repositories[key] = s3obj
+	return s3obj, nil
+}
+
+func (d *s3Driver) Repository(ctx context.Context, server *url.URL, repoName string, insecure bool) (distribution.Repository, error) {
+	parts := strings.SplitN(repoName, "/", 3)
+	if len(parts) < 3 {
+		return nil, fmt.Errorf("you must pass a three segment repository name for s3 uploads, where the first segment is the region and the second segment is the bucket")
+	}
+	s3obj, err := d.newObject(server, parts[0], insecure, &url.URL{Scheme: server.Scheme, Host: server.Host, Path: "/" + repoName})
+	if err != nil {
+		return nil, err
+	}
+	repo := &s3Repository{
+		ctx:      ctx,
+		s3:       s3obj,
+		bucket:   parts[1],
+		repoName: parts[2],
+		copyFrom: d.CopyFrom,
+	}
+	return repo, nil
+}
+
+type s3Repository struct {
+	ctx      context.Context
+	s3       *s3.S3
+	bucket   string
+	once     sync.Once
+	initErr  error
+	copyFrom []string
+
+	repoName string
+}
+
+// Named returns the name of the repository.
+func (r *s3Repository) Named() reference.Named {
+	name, err := reference.ParseNamed(r.repoName)
+	if err != nil {
+		panic(err)
```

Comment:
Can `s3Driver.Repository()` attempt to parse the `repoName` and error out to prevent this panic?

Reply:
Only valid repo names are allowed.  I could pass a named reference I suppose.